src/validation.cpp

```diff
@@ -620,19 +621,58 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
-            // this, see https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-November/016518.html
-            if (nSize >  EXTRA_DESCENDANT_TX_SIZE_LIMIT ||
-                    !pool.CalculateMemPoolAncestors(entry, setAncestors, 2, nLimitAncestorSize, nLimitDescendants + 1, nLimitDescendantSize + EXTRA_DESCENDANT_TX_SIZE_LIMIT, dummy_err_string)) {
-                return state.Invalid(ValidationInvalidReason::TX_MEMPOOL_POLICY, false, REJECT_NONSTANDARD, "too-long-mempool-chain", errString);
+            if (setConflicts.size() == 1) {
+                // If we are an RBF transaction which conflicts directly with exactly one other transaction (but may
+                // evict children of said transaction), conservatively accept transactions which are not adding any
+                // new mempool dependencies (checked first).
+                // Such transactions are clearly not merging any existing packages, so we are only concerned with
+                // ensuring that (a) no package is growing past the package size (not count) limits and (b) we are
+                // not allowing something to effectively use the (below) carve-out spot when it shouldn't be allowed
+                // to.
+                // To check these we re-run CalculateMemPoolAncestors up to twice, once corresponding to the above,
+                // regular, case, and one corresponding to the below, carve-out, case. Each has the same ancestor
+                // count limits as the corresponding case, with the descendant limits increased by the size (and one
+                // count) of the directly-conflicting transaction.
+                assert(setIterConflicting.size() == 1);
+                CTxMemPool::txiter conflict = *setIterConflicting.begin();
+
+                std::set<uint256> conflict_existing_mempool_inputs;
+                for (const CTxIn& input : conflict->GetTx().vin) {
+                    if (mempool.exists(input.prevout.hash)) {
+                        conflict_existing_mempool_inputs.insert(input.prevout.hash);
+                    }
+                }
+
+                for (const CTxIn& new_input: tx.vin) {
+                    if (mempool.exists(new_input.prevout.hash)) {
+                        if (conflict_existing_mempool_inputs.count(new_input.prevout.hash) == 0) {
+                            return state.Invalid(ValidationInvalidReason::TX_MEMPOOL_POLICY, false, REJECT_NONSTANDARD, "too-long-mempool-chain", dummy_err_string);
+                        }
+                    }
+                }
+                if (!pool.CalculateMemPoolAncestors(entry, setAncestors, nLimitAncestors, nLimitAncestorSize, nLimitDescendants + 1,
+                        nLimitDescendantSize + conflict->GetTxSize(), dummy_err_string)) {
+                    if (nSize >  EXTRA_DESCENDANT_TX_SIZE_LIMIT ||
+                            !pool.CalculateMemPoolAncestors(entry, setAncestors, 2, nLimitAncestorSize, nLimitDescendants + 2,
```

Comment:
It looks like you're not clearing setAncestors before this call.  I think that's actually fine, but maybe better practice to clear it out to avoid confusion?